src/chapter19/exercises/exercise05-partf.lean

```diff
@@ -0,0 +1,169 @@
+import tactic
+import combinatorics.pigeonhole
+import data.int.parity
+import data.nat.factorization.basic
+
+
+lemma nat.ord_compl_eq_dvd (a b : ℕ) (h : ord_compl[2] a = ord_compl[2] b) (ha : 0 < a) (hab : a < b) :
+  a ∣ b :=
+begin
+  -- if a = 2^k1 * p, b = 2^k2 * p
+  set k1 : ℕ := a.factorization 2,
+  set k2 : ℕ := b.factorization 2,
+  rw dvd_iff_exists_eq_mul_left,
+  -- c = 2^ (k2 - k1)
+  use (2 ^ (k2 - k1)),
+  have h02 : 0 < 2 := by norm_num,
+  -- because of natural division is involved, we need divisibility
+  have had := nat.ord_proj_dvd a 2,
+  have hbd := nat.ord_proj_dvd b 2,
+  have haf := pow_pos h02 k1,
+  have hbf := pow_pos h02 k2,
+  have hab : k1 ≤ k2,
+  { by_contra hc,
+    push_neg at hc,
+    have hc' : 2 ^ k2 < 2 ^ k1,
+    { rw pow_lt_pow_iff,
+      exact hc,
+      norm_num, },
+    have hak : 0 < a / 2 ^ k1,
+    { apply nat.div_pos,
+      apply nat.ord_proj_le,
+      exact ne_of_gt ha,
+      exact haf, },
+    suffices : b < a,
+    { linarith, },
+    { have hc'' :  2 ^ k2 * (b / 2 ^ k2) < 2 ^ k1 * (a / 2 ^ k1),
+      { rw ← h, apply mul_lt_mul_of_pos_right hc' hak, },
+    rw mul_comm (2 ^ k2) (b / 2 ^ k2) at hc'',
+    rw nat.div_mul_cancel at hc'',
+    swap, exact hbd,
+    rw mul_comm (2 ^ k1) (a / 2 ^ k1) at hc'',
+    rw nat.div_mul_cancel at hc'',
+    swap, exact had,
+    exact hc'', }, },
+  have := nat.pow_div hab h02,
+  rw ← this,
+  -- again we need divisibility to proceed
+  have hkd := pow_dvd_pow 2 hab,
+  rw mul_comm,
+  rw ← nat.mul_div_assoc _ hkd,
+  rw mul_comm a (2^k2),
+  rw nat.mul_div_assoc,
+  swap, exact had,
+  rw h,
+  rw mul_comm,
+  rw nat.div_mul_cancel hbd,
+end
+
+lemma partf (T : finset ℕ) (hT : ∀ t ∈ T, (1 : ℤ) ≤ t ∧ t ≤ 200) (hTcard : T.card = 101) : ∃ a b : ℕ,
+  a ∈ T ∧ b ∈ T ∧ a ≠ b ∧ a ∣ b :=
+begin
+  -- claim : every t can be written as 2^k * q for which q is odd, using ord_compl[2] t
+  let Q : finset ℕ := (finset.Icc 1 200).filter odd,
+  have hQcard : Q.card = 100,
+  { -- Show that it equals (finset.Iio 100).map \<\la n, 2 * n + 1, proof_of_injectivity_here\> 
+    have hQ : Q = (finset.Iio 100).map ⟨λ n, 2 * n + 1 , 
+                                        begin intros a b hab, 
+                                        simpa [add_left_inj, mul_eq_mul_left_iff, bit0_eq_zero, nat.one_ne_zero, or_false] using hab, 
+                                        end ⟩,
+    { dsimp [Q],
+      rw le_antisymm_iff,
+      split,
+      { intros x hx,
+        simp only [finset.mem_map, finset.mem_Iio, function.embedding.coe_fn_mk, exists_prop, nat.one_le_cast, finset.mem_filter,
+          finset.mem_Icc, nat.odd_iff_not_even] at hx ⊢,
+        refine ⟨((x-1)/2), _, _⟩,
+        { rcases hx with ⟨⟨h1, h2⟩, h3⟩,
+          zify at h2 ⊢,
+          rw int.div_lt_iff_lt_mul,
+          linarith,
+          norm_num, },
+        { rcases hx with ⟨⟨h1, h2⟩, h3⟩,
+          rw ← nat.odd_iff_not_even at h3,
+          unfold odd at h3,
```

Comment:
```suggestion
```
You don't need this `unfold` line, you're going to do cases on `h3` and `cases` can see through definitional equality.